### PR TITLE
chore(storybook): Upgrade Storybook from 6 to 8

### LINKED ISSUE #52367: [UI Components] Storybook fails to start with TypeScript 5
### Describe the Issue

I got the following error when I started Storybook:

Screenshot 2023-11-15 at 23 26 40

I was able to start Storybook successfully by downgrading TypeScript to v4.9.5. I think the issue here is our Storybook version isn't compatible with TypeScript 5 (we are using v6 while the latest version is v7), and the resolution should just be upgrading Storybook to v7.

I did try upgrading Storybook a couple weeks ago using a migration command that the library provides, and it was successful. However, since this is a major version bump, there are a handful of deprecated/unsupported parts that we need to throw away. I didn't want to open a PR before I could document all the changes, so the upgrade hasn't happened yet.

I can live with switching between TS 4 and 5 for now, and will make the upgrade, uhm, soon.

### PR DESCRIPTION
**Pull request by valentinpalkovic**
Checklist:

- [x] I have read and followed the [contribution guidelines](https://contribute.freecodecamp.org).
- [x] I have read and followed the [how to open a pull request guide](https://contribute.freecodecamp.org/#/how-to-open-a-pull-request).
- [x] My pull request targets the `main` branch of freeCodeCamp.
- [x] I have tested these changes locally on my machine, or GitPod.

Closes #52367

## Upgrade Storybook from 6 to 8

As one of the core maintainers of Storybook, I have randomly picked your Storybook to upgrade from Storybook 6 to 8 (prerelease). We do this regularly to test different projects and ensure the upgrade and migration experience is smooth and painless.

Feel free to adjust this PR, merge, or close it.

Storybook starts, and the stories and MDX pages are accessible. I don't guarantee that every Story is working as expected, though. Further QA may be necessary. Please reach out if you have any questions!

### CODE REVIEW

**Comment by valentinpalkovic** [comment]
Hi @huyenltnguyen 

I have discovered some missing TypeScript related pieces, which I will handle tomorrow and push them. Afterwards, I let you know in detail what I did :) 

Please keep in mind, that during the beta phase, we are still actively working on the migration experience and things can change. Nevertheless, I will let you k ow precisely what I did.

**Comment by valentinpalkovic** [comment]
Usually, when people migrate, they should use the `npx storybook@next upgrade` command. I did this in a modified way, since we have some upgrade blockers in place, which we have to remove first.

In the second step, I manually removed `@storybook/addon-postcss` and replaced it with `@storybook/addon-styling-web pack.` We will write an automigration to notify the user, that `@storybook/addon-postcss` is deprecated and shouldn't be used anymore.

During the migration, the user will be asked which compiler to use. Since Storybook 8, the Webpack5 builder [doesn't contain a default compiler anymore](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#removed-babelcore-and-babel-loader-from-storybookbuilder-webpack5). The user has to select between `@storybook/addon-webpack5-compiler-babel` or `@storybook/addon-webpack5-compiler-swc` to setup either Babel or SWC as a compiler for Webpack.

Packages that were deprecated in Storybook 7 [have now been removed](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md#removed-deprecated-shim-packages). For example, `@storybook/addons`, which was used in `.storybook/manager.js` to configure the theme. It got replaced by [@storybook/manager-api](https://github.com/freeCodeCamp/freeCodeCamp/pull/53809/files#diff-a088a1449be72ecf7038f4550a0d26c472ce5482821fc7e7fe7df91a63eb99adR1).

Despite that, I was running the following codemods:

- mdx-to-csf
- csf-2-to-3
- upgrade-deprecated-types

They can be run by executing the `migrate` cli:

- `npx storybook@next migrate <codemod> --glob="**/*.stories.tsx"

In the end, I made some manual adjustments to use `StoryObj` instead of `StoryFn` type so that the Stories are written in CSF3 format, instead of 1/2.

**Review by huyenltnguyen — COMMENTED** [review]
@valentinpalkovic Thanks a lot for the work and the breakdown!

There is just a concern from me, but the PR looks great otherwise.

I've also tested the changes locally and verified that:
- Storybook starts successfully (and instantly!)
- The stories are loaded successfully (there are some small UI changes from Storybook, but I can find the release notes and catch up later)

**Note:** Components such as `Col` and `Container` have some display issue in dark theme. This is an existing issue and can be reproduced before the upgrade. I can look into them later.

<details>
  <summary>Screenshot of the issue</summary>

  | Light | Dark |
  | --- | --- |
  | Screenshot 2024-02-23 at 21 37 48 | Screenshot 2024-02-23 at 21 37 38 |

</details>

**Review by huyenltnguyen — APPROVED** [review]

**Review by raisedadead — APPROVED** [review]

File: tools/ui-components/.storybook/preview.js [hunk 1]
```diff
@@ -1,8 +1,11 @@
 /* the styled-elements and normalized are included here to replicate the presets that exist in the learn app */
+import React from 'react';
```
[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply tools/ui-components/.storybook/preview.js:2]: So you are using jsx in this file and that was the reason why it was required :)

File: tools/ui-components/.storybook/preview.js [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
 import '../src/normalize.css';
 import '../src/global-element-styles.css';
 import '../src/base.css';
 
+import '../../../client/src/components/layouts/global.css';
```
[the remark this answers is not in the document]
Comment (valentinpalkovic) [reply tools/ui-components/.storybook/preview.js]: Honestly, this was I think added while adding the new @storybook/addon-styling-webpack package. It asked me for the global css file and I assumed this is it! So just remove it if nothing changes :)

**Pull request merged**